src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs

```diff
@@ -98,9 +98,19 @@ public void CheckTypesOnDeclaration(CheckTypesContext context, DType actualBodyR
             Contracts.AssertValue(actualBodyReturnType);
             Contracts.AssertValue(errorContainer);
 
-            if (!ReturnType.Kind.Equals(actualBodyReturnType.Kind) && !actualBodyReturnType.CoercesTo(ReturnType, true, false, context.Features.PowerFxV1CompatibilityRules))
+            if (!ReturnType.Kind.Equals(actualBodyReturnType.Kind))
             {
-                errorContainer.EnsureError(DocumentErrorSeverity.Severe, UdfBody, TexlStrings.ErrUDF_ReturnTypeDoesNotMatch);
+                if (actualBodyReturnType.CoercesTo(ReturnType, true, false, context.Features.PowerFxV1CompatibilityRules))
+                {
+                    if (!_binding.CoercedToplevelType.IsValid)
```

Comment:
>if (!_binding.CoercedToplevelType.IsValid) [](http://example.com/codeflow?start=20&length=42)

Why do you need this check?

Reply:
Just being defensive here, as have an assertion in SetCoercedToplevelType()

Reply:
Why would we only need to set it if it's not valid? when would it have been set otherwise?

Reply:
Wouldn't the assert failing be the correct behavior?